collector/lib/ConfigLoader.cpp

```diff
@@ -52,41 +56,50 @@ bool ConfigLoader::LoadConfiguration(CollectorConfig& config) {
   return LoadConfiguration(config, node);
 }
 
-bool ConfigLoader::LoadConfiguration(CollectorConfig& config, const YAML::Node& node) {
-  const auto& config_file = CONFIG_FILE.value();
+Json::Value yamlNodeToJson(const YAML::Node& yamlNode) {
+  Json::Value jsonValue;
 
-  if (node.IsNull() || !node.IsDefined() || !node.IsMap()) {
-    CLOG(ERROR) << "Unable to read config from " << config_file;
-    return false;
+  switch (yamlNode.Type()) {
+    case YAML::NodeType::Null:
+      jsonValue = Json::Value::null;
+      break;
+    case YAML::NodeType::Scalar:
+      jsonValue = yamlNode.as<std::string>();
+      break;
+    case YAML::NodeType::Sequence:
+      for (std::size_t i = 0; i < yamlNode.size(); ++i) {
+        jsonValue.append(yamlNodeToJson(yamlNode[i]));
+      }
+      break;
+    case YAML::NodeType::Map:
+      for (const auto& pair : yamlNode) {
+        jsonValue[pair.first.as<std::string>()] = yamlNodeToJson(pair.second);
+      }
+      break;
+    case YAML::NodeType::Undefined:
+    default:
+      break;
   }
 
-  YAML::Node networking_node = node["networking"];
-  if (!networking_node || networking_node.IsNull()) {
-    CLOG(DEBUG) << "No networking in " << config_file;
-    return true;
-  }
+  return jsonValue;
+}
 
-  YAML::Node external_ips_node = networking_node["externalIps"];
-  if (!external_ips_node) {
-    CLOG(DEBUG) << "No external IPs in " << config_file;
-    return true;
+bool ConfigLoader::LoadConfiguration(CollectorConfig& config, const YAML::Node& node) {
+  const auto jsonConfig = yamlNodeToJson(node);
+  Json::StreamWriterBuilder writer;
+  std::string jsonStr = Json::writeString(writer, jsonConfig);
+
+  sensor::CollectorConfig runtimeConfig;
+  google::protobuf::util::JsonParseOptions parseOptions;
+  parseOptions.ignore_unknown_fields = true;
+  auto status = google::protobuf::util::JsonStringToMessage(jsonStr, &runtimeConfig, parseOptions);
```

Comment:
It seems silly to first convert to json, then json string, and then to the protobuf class, instead of converting more directly, and it likely is silly. The reason I did this is that this way I don't have to worry as much about the types. E.g. when converting from yaml, in this case, the code assumes that all the values are strings. This is later fixed by JsonStringToMessage. However, if we tried to convert to the protobuf class directly or the json class used by protobuf, we would first have to figure out the type before doing the conversion.

Reply:
Add comments to explain this.